File: dotnet/Allmark/RenderConsole/ConsoleCodeFenceRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleCodeFenceRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "code_fence",
            Render = Render,
        };
    }

    public static void Render(MarkdownNode node, RendererState state, bool? decode = true)
    {
        var content = (node.Content ?? "").TrimEnd('\n');
        var lines = content.Split('\n');

        if (lines.Length == 0 || (lines.Length == 1 && string.IsNullOrEmpty(lines[0])))
        {
            state.Output.Append($"{Ansi.Dim}┌─{Ansi.Reset}\n{Ansi.Dim}└─{Ansi.Reset}\n");
        }
        else
        {
            state.Output.Append($"{Ansi.Dim}┌─{Ansi.Reset}\n");
            for (int i = 0; i < lines.Length; i++)
            {
                // Skip last empty line
                if (i == lines.Length - 1 && string.IsNullOrEmpty(lines[i]))
                {
                    continue;
                }
                state.Output.Append($"{Ansi.Dim}│{Ansi.Reset} {lines[i].TrimEnd('\r')}\n");
            }
            state.Output.Append($"{Ansi.Dim}└─{Ansi.Reset}\n");
        }
    }
}

File: zig/src/render-console/consoleCodeFenceRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const ConsoleRendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const ansiDim = @import("./renderToConsole.zig").ansiDim;
const ansiReset = @import("./renderToConsole.zig").ansiReset;
const renderChildrenConsole = @import("./renderToConsole.zig").renderChildrenConsole;

pub const consoleCodeFenceRenderer = Renderer{
    .name = "code_fence",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *ConsoleRendererState, first: ?bool, last: ?bool, decode: ?bool) void {
    _ = first;
    _ = last;
    _ = decode;

    const content = std.mem.trimEnd(u8, node.content, "\n");
    const isEmpty = content.len == 0;

    if (isEmpty) {
        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.appendSlice(state.allocator, "┌─") catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
        state.output.append(state.allocator, '\n') catch unreachable;
        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.appendSlice(state.allocator, "└─") catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
    } else {
        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.appendSlice(state.allocator, "┌─") catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
        state.output.append(state.allocator, '\n') catch unreachable;

        var lineIter = std.mem.splitScalar(u8, content, '\n');
        while (lineIter.next()) |line| {
            // Skip last empty line
            if (line.len == 0) {
                const isLast = lineIter.peek() == null;
                if (isLast) {
                    continue;
                }
            }
            state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
            state.output.appendSlice(state.allocator, "│") catch unreachable;
            state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
            state.output.append(state.allocator, ' ') catch unreachable;

            // Trim carriage returns
            const trimmedLine = std.mem.trim(u8, line, "\r");
            state.output.appendSlice(state.allocator, trimmedLine) catch unreachable;
            state.output.append(state.allocator, '\n') catch unreachable;
        }

        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.appendSlice(state.allocator, "└─") catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
    }
}
